File: temporal-research/07_provisioning/activities.py
import asyncio
import logging
import random
import uuid
from typing import Optional

from temporalio import activity

from models import (
    CacheResult,
    DatabaseResult,
    DnsRecord,
    LoadBalancerResult,
    ServerResult,
    SubnetResult,
    VpcResult,
)

logger = logging.getLogger(__name__)


def _id(prefix: str) -> str:
    return f"{prefix}-{uuid.uuid4().hex[:8]}"


@activity.defn
async def provision_vpc(region: str) -> VpcResult:
    logger.info("Provisioning VPC in %s", region)
    await asyncio.sleep(0.4)
    return VpcResult(vpc_id=_id("vpc"), cidr="10.0.0.0/16")


@activity.defn
async def provision_subnet(vpc_id: str, az: str, cidr: str) -> SubnetResult:
    logger.info("Provisioning subnet %s in %s", cidr, az)
    await asyncio.sleep(0.3)
    return SubnetResult(subnet_id=_id("subnet"), availability_zone=az, cidr=cidr)


@activity.defn
async def provision_server(
    vpc_id: str, subnet_id: str, instance_type: str, az: str, environment: str
) -> ServerResult:
    logger.info("Provisioning %s server in %s", instance_type, az)
    await asyncio.sleep(random.uniform(0.8, 1.5))  # servers take variable time
    return ServerResult(
        server_id=_id("i"),
        private_ip=f"10.0.{random.randint(1, 254)}.{random.randint(2, 254)}",
        availability_zone=az,
        instance_type=instance_type,
    )


@activity.defn
async def provision_database(
    vpc_id: str, subnet_ids: list[str], instance_type: str, environment: str
) -> DatabaseResult:
    logger.info("Provisioning RDS (%s) in %s", instance_type, vpc_id)
    await asyncio.sleep(1.5)  # databases take longer
    db_id = _id("db")
    return DatabaseResult(db_id=db_id, endpoint=f"{db_id}.cluster.rds.amazonaws.com")


@activity.defn
async def provision_cache(
    vpc_id: str, subnet_id: str, environment: str
) -> CacheResult:
    logger.info("Provisioning ElastiCache in %s", vpc_id)
    await asyncio.sleep(1.0)
    cache_id = _id("cache")
    return CacheResult(cache_id=cache_id, endpoint=f"{cache_id}.cache.amazonaws.com")


@activity.defn
async def provision_load_balancer(
    vpc_id: str, subnet_ids: list[str], server_ids: list[str], environment: str
) -> LoadBalancerResult:
    logger.info("Provisioning ALB for %d servers", len(server_ids))
    await asyncio.sleep(0.8)
    lb_id = _id("alb")
    return LoadBalancerResult(
        lb_id=lb_id, dns_name=f"{environment}.{lb_id}.elb.amazonaws.com"
    )


@activity.defn
async def register_dns_records(
    records: list[DnsRecord], environment: str
) -> list[DnsRecord]:
    logger.info("Registering %d DNS records for %s", len(records), environment)
    await asyncio.sleep(0.3)
    return records


@activity.defn
async def run_health_check(resource_id: str, endpoint: str) -> dict:
    logger.info("Health check: %s @ %s", resource_id, endpoint)
    await asyncio.sleep(random.uniform(0.1, 0.3))
    return {"resource_id": resource_id, "endpoint": endpoint, "healthy": True}

File: temporal-research-go/07_provisioning/activities.go
package provisioning

import (
	"context"
	"fmt"
	"strings"
	"time"
)

// ProvisionVPC provisions a VPC and returns its ID and CIDR block.
func ProvisionVPC(ctx context.Context, req ProvisionRequest) (VPCResult, error) {
	time.Sleep(500 * time.Millisecond)
	return VPCResult{
		VPCID: fmt.Sprintf("vpc-%s", req.Env),
		CIDR:  "10.0.0.0/16",
	}, nil
}

// ProvisionSubnet provisions a subnet within the given VPC in the given availability zone.
func ProvisionSubnet(ctx context.Context, vpcID, az string) (SubnetResult, error) {
	time.Sleep(300 * time.Millisecond)
	// Derive a short suffix from the AZ, e.g. "us-east-1a" → "1a"
	parts := strings.Split(az, "-")
	suffix := parts[len(parts)-1]
	return SubnetResult{
		SubnetID: fmt.Sprintf("subnet-%s-%s", vpcID, suffix),
		AZ:       az,
	}, nil
}

// ProvisionServer provisions a server in the given subnet.
func ProvisionServer(ctx context.Context, subnetID, az string, idx int) (ServerResult, error) {
	time.Sleep(800 * time.Millisecond)
	return ServerResult{
		ServerID: fmt.Sprintf("srv-%s-%d", subnetID, idx),
		IP:       fmt.Sprintf("10.0.%d.%d", idx, 10+idx),
	}, nil
}

// ProvisionDatabase provisions a managed database in the given VPC.
func ProvisionDatabase(ctx context.Context, vpcID string) (DatabaseResult, error) {
	time.Sleep(1200 * time.Millisecond)
	return DatabaseResult{
		Endpoint: fmt.Sprintf("db.%s.internal", vpcID),
		Port:     5432,
	}, nil
}

// ProvisionCache provisions an in-memory cache cluster in the given VPC.
func ProvisionCache(ctx context.Context, vpcID string) (CacheResult, error) {
	time.Sleep(600 * time.Millisecond)
	return CacheResult{
		Endpoint: fmt.Sprintf("cache.%s.internal", vpcID),
		Port:     6379,
	}, nil
}

// ProvisionLoadBalancer provisions a load balancer targeting the given server IDs.
func ProvisionLoadBalancer(ctx context.Context, vpcID string, serverIDs []string) (LBResult, error) {
	time.Sleep(400 * time.Millisecond)
	return LBResult{
		DNS: fmt.Sprintf("lb.%s.example.com", vpcID),
	}, nil
}

// RegisterDNS creates a DNS record pointing the environment's hostname to the target.
func RegisterDNS(ctx context.Context, env, target string) (DNSRecord, error) {
	time.Sleep(200 * time.Millisecond)
	return DNSRecord{
		Name:   fmt.Sprintf("%s.example.com", env),
		Target: target,
	}, nil
}

// RunHealthCheck verifies that the given target is healthy.
func RunHealthCheck(ctx context.Context, target string) (bool, error) {
	time.Sleep(300 * time.Millisecond)
	return true, nil
}
